rust-webvr/src/api/vrexternal/display.rs

```diff
@@ -0,0 +1,265 @@
+use super::{mozgfx, VRExternalShmemPtr};
+use rust_webvr_api::utils;
+use std::cell::RefCell;
+use std::sync::Arc;
+use std::thread;
+use std::time::Duration;
+use {
+    VRDisplay, VRDisplayData, VRDisplayPtr, VRFrameData, VRFramebuffer, VRFramebufferAttributes,
+    VRLayer, VRViewport,
+};
+
+pub struct VRExternalDisplay {
+    last_sensor_frame_id: u64,
+    rendered_layer: Option<VRLayer>,
+    shmem: VRExternalShmemPtr,
+    display_id: u32,
+    attributes: VRFramebufferAttributes,
+    presenting: bool,
+}
+
+impl VRExternalDisplay {
+    pub fn new(shmem: VRExternalShmemPtr) -> VRDisplayPtr {
+        let id = shmem.system().as_ref().sensorState.inputFrameID;
+        Arc::new(RefCell::new(VRExternalDisplay {
+            last_sensor_frame_id: id,
+            rendered_layer: None,
+            shmem,
+            display_id: utils::new_id(),
+            attributes: Default::default(),
+            presenting: false,
+        }))
+    }
+}
+
+impl VRExternalDisplay {
+    fn block_until_sensor_state_changed(&mut self) {
+        loop {
+            {
+                let sys = self.shmem.system();
+                let id = sys.as_ref().sensorState.inputFrameID;
+                let _gen = sys.as_ref().displayState.mPresentingGeneration;
+                // FIXME: Note for reviewer: what to do if mPresentingGeneration changed?
+                let prev_id = self.last_sensor_frame_id;
+                if id != prev_id {
```

Comment:
We should handle some extra situations to exit the wait loop. See how this wait is implemented in Gecko

```
  // PullState blocks until we have a new id, the device is disconnected or mSuppressFrames has been set.
  PullState([&]() {
    return (mDisplayInfo.mDisplayState.mLastSubmittedFrameId >= aFrameId) ||
            mDisplayInfo.mDisplayState.mSuppressFrames ||
            !mDisplayInfo.mDisplayState.mIsConnected;
  });

  if (mDisplayInfo.mDisplayState.mSuppressFrames || !mDisplayInfo.mDisplayState.mIsConnected) {
    // External implementation wants to supress frames, service has shut down or hardware has been disconnected.
    return false;
  }
```

Reply:
Note, only the `mPresentingGeneration` change should trigger a servo WebVR session exit. The other situations (mSuppressFrames and disconnect) can last just one or a few frames so are recoverable